lite/verifying/provider.go

```diff
@@ -0,0 +1,440 @@
+// not error like that, so that e.g. cli usage never errors unexpectedly.
+func NewProvider(chainID, rootDir string, client lclient.SignStatusClient, logger log.Logger, cacheSize int, options TrustOptions) (*Provider, error) {
+
+	vp := initProvider(chainID, rootDir, client, logger, cacheSize, options)
+
+	// Get the latest source commit, or the one provided in options.
+	trustCommit, err := getTrustCommit(client, options)
+	if err != nil {
+		return nil, err
+	}
+
+	err = vp.fillValsetAndSaveFC(trustCommit, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// sanity check
+	if time.Now().Sub(trustCommit.Time) <= 0 {
+		panic(fmt.Sprintf("impossible time %v vs %v", time.Now(), trustCommit.Time))
+	}
+
+	// Otherwise we're syncing within the unbonding period.
+	// NOTE: There is a duplication of fetching this latest commit (since
+	// UpdateToHeight() will fetch it again, and latestCommit isn't used), but
+	// it's only once upon initialization of a validator so it's not a big
+	// deal.
+	if options.TrustHeight > 0 {
+		latestCommit, err := client.Commit(nil)
+		if err != nil {
+			return nil, err
+		}
+		err = vp.UpdateToHeight(chainID, latestCommit.SignedHeader.Height)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return vp, nil
+}
+
+// getTrustCommit returns a commit trusted with weak subjectivity. It either:
+// 1. Fetches a commit at height provided in options and ensures the specified commit
+// 	is within the trust period of latest block
+// 2. Trusts the remote node and gets the latest commit
+// 3. Returns an error if the height provided in trust option is too old to sync to latest.
+func getTrustCommit(client lclient.SignStatusClient, options TrustOptions) (types.SignedHeader, error) {
+
+	// Get the lastest commit always
+	latestBlock, err := client.Commit(nil)
+	if err != nil {
+		return types.SignedHeader{}, err
+	}
+
+	// If the user has set a root of trust, confirm it then update to newest
+	if options.TrustHeight != 0 && options.TrustHash != nil {
+		trustBlock, err := client.Commit(&options.TrustHeight)
+		if err != nil {
+			return types.SignedHeader{}, err
+		}
+
+		if latestBlock.Time.Sub(trustBlock.Time) > options.TrustPeriod {
+			return types.SignedHeader{}, fmt.Errorf("Your Trusted Block Height is older than the trust period from Latest Block")
+		}
+
+		trustCommit := trustBlock.SignedHeader
+		if !bytes.Equal(trustCommit.Hash(), options.TrustHash) {
+			return types.SignedHeader{}, fmt.Errorf("WARNING!!! Expected height/hash %v/%X but got %X",
+				options.TrustHeight, options.TrustHash, trustCommit.Hash())
+		}
+		return trustCommit, nil
+	} else {
+
+		latestCommit := latestBlock.SignedHeader
+
+		// NOTE: This should really belong in the callback.
+		// WARN THE USER IN ALL CAPS THAT THE LITE CLIENT IS NEW,
+		// AND THAT WE WILL SYNC TO AND VERIFY LATEST COMMIT.
+		fmt.Printf("trusting source at height %v and hash %X...\n", latestCommit.Height, latestCommit.Hash())
```

Comment:
?

Reply:
I think this is to warn users that if the lite client starts afresh (it does not trust anything yet), it will only fetch the latest commit (and not verify from the first height). Is that right? @zmanian @jackzampolin

Reply:
that was the intent here!

Reply:
Can we explicitly require user input here?

Reply:
Shouldn't we require user input outside of this package? I don't this package needs to think about user input.